File: apps/passenger-android/src/main/kotlin/lk/mageride/passenger/map/VehicleLayers.kt
package lk.mageride.passenger.map

import lk.mageride.shared.data.models.VehicleType
import org.maplibre.android.maps.Style
import org.maplibre.android.style.expressions.Expression
import org.maplibre.android.style.layers.CircleLayer
import org.maplibre.android.style.layers.LineLayer
import org.maplibre.android.style.layers.PropertyFactory
import org.maplibre.android.style.layers.SymbolLayer
import org.maplibre.android.style.sources.GeoJsonOptions
import org.maplibre.android.style.sources.GeoJsonSource

/**
 * The MAP-* primitives D2' §0.3 makes a hard rule, installed onto a style.
 *
 * > *"MapLibre GL Native style (light/dark PMTiles), `SymbolLayer` markers by `vehVeh*` color +
 * > heading arrow (MAP-06), `ClusterLayer` when zoomed out (MAP-05), interpolated marker animation
 * > (MAP-04), `LineLayer` trip polyline (MAP-08), accuracy circle (MAP-02), 100m geofence circle
 * > (MAP-10). Pins: `pickup` green, `dropoff` red, `user` blue dot."*
 *
 * Everything on that list except MAP-04 is a layer, and every one of them is here; MAP-04 is
 * [MarkerInterpolator], because a tween is arithmetic rather than cartography and MapLibre does
 * not tween a source's contents. MAP-09 — offline tile caching — is the one item of the ten this
 * component does not carry; see [MapStyles].
 *
 * This owns the sources and the layers, not their contents: what goes *in* a source is
 * [MageRideMap]'s and, above it, the screen's.
 *
 * **Colours are passed in, never written here.** They come from the theme's
 * [lk.mageride.passenger.ui.theme.VehicleColors] through [markerColourExpression]; a hex in this
 * file would be the second copy of §0.2's table.
 */
internal object VehicleLayers {

    /** Source id for live vehicle positions. Features carry [PROP_VEHICLE_TYPE] and [PROP_HEADING]. */
    const val SOURCE_VEHICLES: String = "mageride-vehicles"

    /** Source id for MAP-02's accuracy circle and MAP-10's 100 m geofence. */
    const val SOURCE_CIRCLES: String = "mageride-circles"

    /** Source id for MAP-08's trip polyline. */
    const val SOURCE_ROUTE: String = "mageride-route"

    /** Source id for §0.3's pickup / dropoff / user pins. */
    const val SOURCE_PINS: String = "mageride-pins"

    const val LAYER_VEHICLES: String = "mageride-vehicles-symbols"
    const val LAYER_CLUSTERS: String = "mageride-vehicles-clusters"
    const val LAYER_CLUSTER_COUNT: String = "mageride-vehicles-cluster-count"
    const val LAYER_ACCURACY: String = "mageride-accuracy"
    const val LAYER_GEOFENCE: String = "mageride-geofence"
    const val LAYER_ROUTE_CASING: String = "mageride-route-casing"
    const val LAYER_ROUTE: String = "mageride-route-line"
    const val LAYER_PINS: String = "mageride-pins-circles"

    /** Feature property carrying the canonical vehicle type (AL-09), so MAP-03 can colour it. */
    const val PROP_VEHICLE_TYPE: String = "vehicleType"

    /** Feature property carrying the heading in degrees — MAP-06's arrow rotation. */
    const val PROP_HEADING: String = "heading"

    /** Feature property carrying the vehicle id, so MAP-07's tap knows what was tapped. */
    const val PROP_VEHICLE_ID: String = "vehicleId"

    /** Feature property distinguishing the accuracy circle from the geofence one. */
    const val PROP_CIRCLE_KIND: String = "circleKind"

    /** Feature property selecting a pin's colour — one of [PIN_PICKUP], [PIN_DROPOFF], [PIN_USER]. */
    const val PROP_PIN_KIND: String = "pinKind"

    const val CIRCLE_ACCURACY: String = "accuracy"
    const val CIRCLE_GEOFENCE: String = "geofence"

    const val PIN_PICKUP: String = "pickup"
    const val PIN_DROPOFF: String = "dropoff"
    const val PIN_USER: String = "user"

    /** D5' §7's arrival geofence, and the radius MAP-10 draws. */
    const val GEOFENCE_RADIUS_M: Int = 100

    /**
     * Installs every layer, **in draw order**, on a style that has just loaded.
     *
     * The order is the argument for having one function rather than four: MapLibre stacks a layer
     * on top of whatever is already there, so circles first (they are context, and a vehicle drawn
     * under its own accuracy halo disappears), then the trip polyline, then the vehicles, then the
     * pins that mark where the passenger is going. A caller that added them itself would get the
     * order right the first time and wrong the second.
     *
     * @param markerImage The image id already added to the style with `style.addImage(...)` —
     *   MAP-06's heading arrow. [MageRideMap] adds it; nothing here ships an asset.
     * @param palette Every colour, resolved from D2' §0.2 by [mapPalette]. Passed in rather than
     *   read here so this file stays free of Compose and free of hexes.
     */
    fun install(style: Style, markerImage: String, palette: MapPalette) {
        addCircles(style, palette.circles)
        addRoute(style, palette.route)
        addVehicles(style, markerImage, palette.vehicles)
        addPins(style, palette.pins)
    }

    /**
     * MAP-03 / MAP-05 / MAP-06 — the vehicle source and its three layers.
     *
     * **Clustering is on the source, not the layer** — that is how MapLibre expresses MAP-05, and
     * why [LAYER_CLUSTERS] filters on `point_count` while [LAYER_VEHICLES] filters on its absence.
     * Without the paired filters a clustered point renders twice: once as its cluster and once as
     * itself.
     */
    private fun addVehicles(style: Style, markerImage: String, palette: VehiclePalette) {
        style.addSource(
            GeoJsonSource(
                SOURCE_VEHICLES,
                GeoJsonOptions()
                    .withCluster(true)
                    // Below zoom 14 a city's worth of vehicles is an unreadable smear of pins;
                    // above it a passenger is looking at individual vehicles near a junction.
                    .withClusterMaxZoom(CLUSTER_MAX_ZOOM)
                    .withClusterRadius(CLUSTER_RADIUS_PX),
            ),
        )

        style.addLayer(
            SymbolLayer(LAYER_VEHICLES, SOURCE_VEHICLES)
                .withProperties(
                    PropertyFactory.iconImage(markerImage),
                    // MAP-03. The whole eleven-colour legend as one `match` — this map draws every
                    // mode in a cell at once, so a single `icon-color` (which is what the driver's
                    // own-vehicle map uses) cannot express it.
                    PropertyFactory.iconColor(markerColourExpression(palette.legend, palette.fallback)),
                    // MAP-06. `icon-rotate` off the feature's own heading, and
                    // `icon-rotation-alignment: map` so the arrow points where the vehicle is
                    // going rather than where the screen is turned.
                    PropertyFactory.iconRotate(Expression.get(PROP_HEADING)),
                    PropertyFactory.iconRotationAlignment("map"),
                    PropertyFactory.iconAllowOverlap(true),
                    PropertyFactory.iconIgnorePlacement(true),
                )
                .withFilter(Expression.not(Expression.has("point_count"))),
        )

        style.addLayer(
            CircleLayer(LAYER_CLUSTERS, SOURCE_VEHICLES)
                .withProperties(
                    PropertyFactory.circleColor(palette.cluster),
                    PropertyFactory.circleRadius(
                        Expression.step(
                            Expression.get("point_count"),
                            Expression.literal(CLUSTER_RADIUS_SMALL),
                            Expression.stop(CLUSTER_STEP_MEDIUM, CLUSTER_RADIUS_MEDIUM),
                            Expression.stop(CLUSTER_STEP_LARGE, CLUSTER_RADIUS_LARGE),
                        ),
                    ),
                )
                .withFilter(Expression.has("point_count")),
        )

        style.addLayer(
            SymbolLayer(LAYER_CLUSTER_COUNT, SOURCE_VEHICLES)
                .withProperties(
                    PropertyFactory.textField(Expression.toString(Expression.get("point_count"))),
                    PropertyFactory.textSize(CLUSTER_TEXT_SIZE),
                    PropertyFactory.textColor(palette.onCluster),
                    PropertyFactory.textIgnorePlacement(true),
                    PropertyFactory.textAllowOverlap(true),
                )
                .withFilter(Expression.has("point_count")),
        )
    }

    /**
     * MAP-02's accuracy circle and MAP-10's 100 m geofence.
     *
     * One source and two layers rather than two sources: both are circles around a point and
     * differ only in radius and paint, and a feature's [PROP_CIRCLE_KIND] is what selects between
     * them. The radii are in metres, which is what `circleRadius` needs converting for — the
     * caller passes pixels computed from the current zoom, because MapLibre's circle radius is a
     * screen measurement and a metre is not.
     */
    private fun addCircles(style: Style, palette: CirclePalette) {
        style.addSource(GeoJsonSource(SOURCE_CIRCLES))

        style.addLayer(
            CircleLayer(LAYER_ACCURACY, SOURCE_CIRCLES)
                .withProperties(
                    PropertyFactory.circleColor(palette.accuracy),
                    PropertyFactory.circleOpacity(ACCURACY_OPACITY),
                    PropertyFactory.circleStrokeWidth(0f),
                )
                .withFilter(Expression.eq(Expression.get(PROP_CIRCLE_KIND), CIRCLE_ACCURACY)),
        )

        style.addLayer(
            CircleLayer(LAYER_GEOFENCE, SOURCE_CIRCLES)
                .withProperties(
                    PropertyFactory.circleColor(palette.geofence),
                    PropertyFactory.circleOpacity(GEOFENCE_OPACITY),
                    PropertyFactory.circleStrokeColor(palette.geofence),
                    PropertyFactory.circleStrokeWidth(GEOFENCE_STROKE_PX),
                )
                .withFilter(Expression.eq(Expression.get(PROP_CIRCLE_KIND), CIRCLE_GEOFENCE)),
        )
    }

    /**
     * MAP-08 — the trip polyline, as two `LineLayer`s.
     *
     * A casing under the line rather than one stroke: the basemap's own roads are white on grey
     * in the light style, so a single orange line at four pixels reads as a road on a screen in
     * sunlight. The casing is the same geometry, wider and darker, which is how every navigation
     * map draws a route and is the cheapest way to keep it legible over both styles.
     */
    private fun addRoute(style: Style, palette: RoutePalette) {
        style.addSource(GeoJsonSource(SOURCE_ROUTE))

        style.addLayer(
            LineLayer(LAYER_ROUTE_CASING, SOURCE_ROUTE).withProperties(
                PropertyFactory.lineColor(palette.casing),
                PropertyFactory.lineWidth(ROUTE_CASING_WIDTH_PX),
                PropertyFactory.lineCap("round"),
                PropertyFactory.lineJoin("round"),
            ),
        )

        style.addLayer(
            LineLayer(LAYER_ROUTE, SOURCE_ROUTE).withProperties(
                PropertyFactory.lineColor(palette.line),
                PropertyFactory.lineWidth(ROUTE_WIDTH_PX),
                PropertyFactory.lineCap("round"),
                PropertyFactory.lineJoin("round"),
            ),
        )
    }

    /**
     * §0.3's *"`pickup` green, `dropoff` red, `user` blue dot"*.
     *
     * Circles rather than teardrop symbols, because a circle needs no asset and the spec's own
     * word for the third one is "dot". A screen that wants a labelled teardrop for a pickup adds
     * its own `SymbolLayer` above this one; what must not vary between screens is the *colour*,
     * which is why the three are one table passed in from the theme.
     */
    private fun addPins(style: Style, palette: PinPalette) {
        style.addSource(GeoJsonSource(SOURCE_PINS))

        val stops = palette.byKind.flatMap { (kind, colour) ->
            listOf(Expression.literal(kind), Expression.color(colour))
        }.toTypedArray()

        style.addLayer(
            CircleLayer(LAYER_PINS, SOURCE_PINS).withProperties(
                PropertyFactory.circleColor(
                    // Three stops, one per pin kind. `Expression.match` is vararg-only, hence
                    // the spread; the fallback is the user dot's own colour, because an unknown
                    // pin kind is a client bug and drawing nothing would hide it.
                    @Suppress("SpreadOperator")
                    Expression.match(
                        Expression.get(PROP_PIN_KIND),
                        Expression.color(palette.byKind[PIN_USER] ?: palette.stroke),
                        *stops,
                    ),
                ),
                PropertyFactory.circleRadius(PIN_RADIUS_PX),
                PropertyFactory.circleStrokeColor(palette.stroke),
                PropertyFactory.circleStrokeWidth(PIN_STROKE_PX),
            ),
        )
    }

    /**
     * MAP-03's legend as a MapLibre `match` expression over [PROP_VEHICLE_TYPE].
     *
     * Built from the caller's colours rather than from constants here, so the one place a vehicle
     * colour is written stays `ui/theme/Color.kt`. The fallback is `vehPrivate` grey — a vehicle
     * type this build does not know is still drawn, because a missing marker reads as a platform
     * outage and an unfamiliar grey one reads as what it is.
     *
     * @param colours Wire name → ARGB int, from `VehicleColors.Legend`.
     * @param fallback `vehPrivate`.
     */
    // MapLibre's `Expression.match` is vararg-only, so the stops have to be spread. The copy is
    // eleven elements, built once when a style is installed rather than per frame.
    @Suppress("SpreadOperator")
    fun markerColourExpression(colours: Map<String, Int>, fallback: Int): Expression {
        val stops = colours.flatMap { (type, colour) ->
            listOf(Expression.literal(type), Expression.color(colour))
        }.toTypedArray()
        return Expression.match(Expression.get(PROP_VEHICLE_TYPE), Expression.color(fallback), *stops)
    }

    /** The wire spelling of every canonical vehicle type, for building the colour map. */
    fun vehicleTypeNames(): List<String> = VehicleType.entries.map(VehicleType::wire)

    private const val CLUSTER_MAX_ZOOM = 14
    private const val CLUSTER_RADIUS_PX = 60
    private const val CLUSTER_RADIUS_SMALL = 16f
    private const val CLUSTER_RADIUS_MEDIUM = 22f
    private const val CLUSTER_RADIUS_LARGE = 28f
    private const val CLUSTER_STEP_MEDIUM = 10
    private const val CLUSTER_STEP_LARGE = 40
    private const val CLUSTER_TEXT_SIZE = 12f
    private const val ACCURACY_OPACITY = 0.15f
    private const val GEOFENCE_OPACITY = 0.08f
    private const val GEOFENCE_STROKE_PX = 2f
    private const val ROUTE_WIDTH_PX = 5f
    private const val ROUTE_CASING_WIDTH_PX = 9f
    private const val PIN_RADIUS_PX = 7f
    private const val PIN_STROKE_PX = 2f
}

File: apps/passenger-ios/PassengerApp/Map/VehicleLayers.swift
import Foundation
import MapLibre
import UIKit

/// The MAP-* primitives D2' §0.3 makes a hard rule, installed onto a MapLibre style.
///
/// > *"MapLibre GL Native style (light/dark PMTiles), `SymbolLayer` markers by `vehVeh*` color +
/// > heading arrow (MAP-06), `ClusterLayer` when zoomed out (MAP-05), interpolated marker animation
/// > (MAP-04), `LineLayer` trip polyline (MAP-08), accuracy circle (MAP-02), 100m geofence circle
/// > (MAP-10). Pins: `pickup` green, `dropoff` red, `user` blue dot."*
///
/// Everything on that list except MAP-04 is a layer, and every one of them is here; MAP-04 is
/// ``MarkerInterpolator``, because a tween is arithmetic rather than cartography and MapLibre does
/// not tween a source's contents. MAP-09 is the one item this component does not carry — see
/// ``MapStyles``.
///
/// **This owns the sources and the layers, not their contents**: what goes *in* a source is
/// ``MageRideMap``'s and, above it, the screen's.
///
/// **The identifiers and the feature-property names are byte-for-byte the Android ones**, which
/// matters more than it looks: the features both apps build come from the *same* `:shared` DTOs, so
/// a property renamed on one side is a marker that silently stops being coloured on that side only.
///
/// **Colours are passed in, never written here.** They come from ``MapPalette``; a `UIColor(red:…)`
/// in this file would be the second copy of §0.2's table.
enum VehicleLayers {

    // MARK: - Identifiers. Byte-for-byte `apps/passenger-android/.../map/VehicleLayers.kt`'s.

    /// Source id for live vehicle positions. Features carry ``propVehicleType`` and ``propHeading``.
    static let sourceVehicles = "mageride-vehicles"

    /// Source id for MAP-02's accuracy circle and MAP-10's 100 m geofence.
    static let sourceCircles = "mageride-circles"

    /// Source id for MAP-08's trip polyline.
    static let sourceRoute = "mageride-route"

    /// Source id for SCR-PI-009's walk-to-halt line.
    static let sourceWalk = "mageride-walk"

    /// Source id for §0.3's pickup / dropoff / user pins.
    static let sourcePins = "mageride-pins"

    static let layerVehicles = "mageride-vehicles-symbols"
    static let layerClusters = "mageride-vehicles-clusters"
    static let layerClusterCount = "mageride-vehicles-cluster-count"
    static let layerAccuracy = "mageride-accuracy"
    static let layerGeofence = "mageride-geofence"
    static let layerRouteCasing = "mageride-route-casing"
    static let layerRoute = "mageride-route-line"
    static let layerWalk = "mageride-walk-line"
    static let layerPins = "mageride-pins-circles"

    /// Feature property carrying the canonical vehicle type (AL-09), so MAP-03 can colour it.
    static let propVehicleType = "vehicleType"

    /// Feature property carrying the heading in degrees — MAP-06's arrow rotation.
    static let propHeading = "heading"

    /// Feature property carrying the vehicle id, so MAP-07's tap knows what was tapped.
    static let propVehicleId = "vehicleId"

    /// Feature property distinguishing the accuracy circle from the geofence one.
    static let propCircleKind = "circleKind"

    /// Feature property selecting a pin's colour.
    static let propPinKind = "pinKind"

    static let circleAccuracy = "accuracy"
    static let circleGeofence = "geofence"

    static let pinPickup = "pickup"
    static let pinDropoff = "dropoff"
    static let pinUser = "user"

    /// D5' §7's arrival geofence, and the radius MAP-10 draws.
    static let geofenceRadiusMetres: Double = 100

    /// The style image id the vehicle symbol layer draws — MAP-06's heading arrow.
    static let markerImage = "mageride-vehicle"

    // MARK: - Installation

    /// Installs every layer, **in draw order**, on a style that has just loaded.
    ///
    /// The order is the argument for having one function rather than five: MapLibre stacks a layer
    /// on top of whatever is already there, so circles first (they are context, and a vehicle drawn
    /// under its own accuracy halo disappears), then the trip polyline and the walking leg, then the
    /// vehicles, then the pins that mark where the passenger is going. A caller that added them
    /// itself would get the order right the first time and wrong the second.
    ///
    /// - Parameter markerImage: the image already set on the style with `style.setImage(_:forName:)`.
    ///   ``MageRideMap`` renders it from ``VehicleToken``; nothing here ships an asset.
    static func install(on style: MLNStyle, markerImage: String, palette: MapPalette) {
        addCircles(to: style, palette: palette.circles)
        addRoute(to: style, palette: palette.route)
        addWalk(to: style, colour: palette.route.walk)
        addVehicles(to: style, markerImage: markerImage, palette: palette.vehicles)
        addPins(to: style, palette: palette.pins)
    }

    /// MAP-03 / MAP-05 / MAP-06 — the vehicle source and its three layers.
    ///
    /// **Clustering is on the source, not the layer** — that is how MapLibre expresses MAP-05, and
    /// why the cluster layers filter on `cluster == YES` while the symbol layer filters on its
    /// absence. Without the paired predicates a clustered point renders twice: once as its cluster
    /// and once as itself.
    private static func addVehicles(to style: MLNStyle, markerImage: String, palette: VehiclePalette) {
        let source = MLNShapeSource(
            identifier: sourceVehicles,
            shape: nil,
            options: [
                .clustered: true,
                // Below zoom 14 a city's worth of vehicles is an unreadable smear of pins; above it
                // a passenger is looking at individual vehicles near a junction.
                .maximumZoomLevelForClustering: clusterMaxZoom,
                .clusterRadius: clusterRadiusPx,
            ]
        )
        style.addSource(source)

        let vehicles = MLNSymbolStyleLayer(identifier: layerVehicles, source: source)
        vehicles.iconImageName = NSExpression(forConstantValue: markerImage)
        // MAP-03. The whole eleven-colour legend as one `match` — this map draws every mode in a
        // cell at once, so a single tint (which is what the driver's own-vehicle map uses) cannot
        // express it.
        vehicles.iconColor = markerColourExpression(palette: palette)
        // MAP-06. Rotation off the feature's own heading, aligned to the map so the arrow points
        // where the vehicle is going rather than where the screen is turned.
        vehicles.iconRotation = NSExpression(forKeyPath: propHeading)
        vehicles.iconRotationAlignment = NSExpression(forConstantValue: "map")
        vehicles.iconAllowsOverlap = NSExpression(forConstantValue: true)
        vehicles.iconIgnoresPlacement = NSExpression(forConstantValue: true)
        vehicles.predicate = NSPredicate(format: "cluster != YES")
        style.addLayer(vehicles)

        let clusters = MLNCircleStyleLayer(identifier: layerClusters, source: source)
        clusters.circleColor = NSExpression(forConstantValue: palette.cluster)
        // The typed initialiser rather than an `NSExpression(format:)` string: MapLibre's expression
        // function names are a compatibility surface inherited from the SDK it forked, and a
        // mistyped one is a runtime exception on the first style load rather than a compile error.
        clusters.circleRadius = NSExpression(
            forMLNStepping: NSExpression(forKeyPath: "point_count"),
            from: NSExpression(forConstantValue: clusterRadiusSmall),
            stops: NSExpression(forConstantValue: [
                clusterStepMedium: clusterRadiusMedium,
                clusterStepLarge: clusterRadiusLarge,
            ])
        )
        clusters.predicate = NSPredicate(format: "cluster == YES")
        style.addLayer(clusters)

        let counts = MLNSymbolStyleLayer(identifier: layerClusterCount, source: source)
        counts.text = NSExpression(format: "CAST(point_count, 'NSString')")
        counts.textFontSize = NSExpression(forConstantValue: clusterTextSize)
        counts.textColor = NSExpression(forConstantValue: palette.onCluster)
        counts.textIgnoresPlacement = NSExpression(forConstantValue: true)
        counts.textAllowsOverlap = NSExpression(forConstantValue: true)
        counts.predicate = NSPredicate(format: "cluster == YES")
        style.addLayer(counts)
    }

    /// MAP-02's accuracy circle and MAP-10's 100 m geofence.
    ///
    /// One source and two layers rather than two sources: both are circles around a point and differ
    /// only in radius and paint, and a feature's ``propCircleKind`` selects between them.
    ///
    /// **The radii are metres and MapLibre's `circleRadius` is pixels**, so both are re-scaled
    /// whenever the camera settles — see ``MageRideMap``. A radius set once is wrong at every other
    /// zoom, which is the whole point of a *100 m* geofence.
    private static func addCircles(to style: MLNStyle, palette: CirclePalette) {
        let source = MLNShapeSource(identifier: sourceCircles, shape: nil, options: nil)
        style.addSource(source)

        let accuracy = MLNCircleStyleLayer(identifier: layerAccuracy, source: source)
        accuracy.circleColor = NSExpression(forConstantValue: palette.accuracy)
        accuracy.circleOpacity = NSExpression(forConstantValue: accuracyOpacity)
        accuracy.circleStrokeWidth = NSExpression(forConstantValue: 0)
        accuracy.predicate = NSPredicate(format: "%K == %@", propCircleKind, circleAccuracy)
        style.addLayer(accuracy)

        let geofence = MLNCircleStyleLayer(identifier: layerGeofence, source: source)
        geofence.circleColor = NSExpression(forConstantValue: palette.geofence)
        geofence.circleOpacity = NSExpression(forConstantValue: geofenceOpacity)
        geofence.circleStrokeColor = NSExpression(forConstantValue: palette.geofence)
        geofence.circleStrokeWidth = NSExpression(forConstantValue: geofenceStrokePx)
        geofence.predicate = NSPredicate(format: "%K == %@", propCircleKind, circleGeofence)
        style.addLayer(geofence)
    }

    /// MAP-08 — the trip polyline, as two line layers.
    ///
    /// A casing under the line rather than one stroke: the basemap's own roads are white on grey in
    /// the light style, so a single orange line at four pixels reads as a road on a screen in
    /// sunlight. The casing is the same geometry, wider and darker, which is how every navigation
    /// map draws a route and is the cheapest way to keep it legible over both styles.
    private static func addRoute(to style: MLNStyle, palette: RoutePalette) {
        let source = MLNShapeSource(identifier: sourceRoute, shape: nil, options: nil)
        style.addSource(source)

        let casing = MLNLineStyleLayer(identifier: layerRouteCasing, source: source)
        casing.lineColor = NSExpression(forConstantValue: palette.casing)
        casing.lineWidth = NSExpression(forConstantValue: routeCasingWidthPx)
        casing.lineCap = NSExpression(forConstantValue: "round")
        casing.lineJoin = NSExpression(forConstantValue: "round")
        style.addLayer(casing)

        let line = MLNLineStyleLayer(identifier: layerRoute, source: source)
        line.lineColor = NSExpression(forConstantValue: palette.line)
        line.lineWidth = NSExpression(forConstantValue: routeWidthPx)
        line.lineCap = NSExpression(forConstantValue: "round")
        line.lineJoin = NSExpression(forConstantValue: "round")
        style.addLayer(line)
    }

    /// SCR-PI-009's *"blue walking polyline … to the closest halt"*.
    ///
    /// **Blue and dashed, and both matter.** The route line is the vehicle's colour and solid; this
    /// one is the passenger's own leg of the journey and is not a road the bus takes. D2'
    /// §SCR-PA-009 spells the treatment out — *"`LineLayer` **blue, dashed**"* — and drawing it in
    /// the same style as the route would say the bus goes to your front door.
    private static func addWalk(to style: MLNStyle, colour: UIColor) {
        let source = MLNShapeSource(identifier: sourceWalk, shape: nil, options: nil)
        style.addSource(source)

        let walk = MLNLineStyleLayer(identifier: layerWalk, source: source)
        walk.lineColor = NSExpression(forConstantValue: colour)
        walk.lineWidth = NSExpression(forConstantValue: walkWidthPx)
        walk.lineCap = NSExpression(forConstantValue: "round")
        // Dash lengths are in line widths, not points, so the pattern scales with the stroke and
        // stays legible at every zoom.
        walk.lineDashPattern = NSExpression(forConstantValue: [walkDash, walkGap])
        style.addLayer(walk)
    }

    /// §0.3's *"`pickup` green, `dropoff` red, `user` blue dot"*.
    ///
    /// Circles rather than teardrop symbols, because a circle needs no asset and the spec's own word
    /// for the third one is "dot". A screen that wants a labelled teardrop for a pickup adds its own
    /// symbol layer above this one; what must not vary between screens is the *colour*, which is why
    /// the three are one table passed in.
    private static func addPins(to style: MLNStyle, palette: PinPalette) {
        let source = MLNShapeSource(identifier: sourcePins, shape: nil, options: nil)
        style.addSource(source)

        var stops: [NSExpression: NSExpression] = [:]
        for (kind, colour) in palette.byKind {
            stops[NSExpression(forConstantValue: kind)] = NSExpression(forConstantValue: colour)
        }

        let pins = MLNCircleStyleLayer(identifier: layerPins, source: source)
        pins.circleColor = NSExpression(
            forMLNMatchingKey: NSExpression(forKeyPath: propPinKind),
            in: stops,
            // The user dot's own colour is the fallback, because an unknown pin kind is a client bug
            // and drawing nothing would hide it.
            default: NSExpression(forConstantValue: palette.byKind[pinUser] ?? palette.stroke)
        )
        pins.circleRadius = NSExpression(forConstantValue: pinRadiusPx)
        pins.circleStrokeColor = NSExpression(forConstantValue: palette.stroke)
        pins.circleStrokeWidth = NSExpression(forConstantValue: pinStrokePx)
        style.addLayer(pins)
    }

    /// MAP-03's legend as an `NSExpression` over ``propVehicleType``.
    ///
    /// The fallback is `vehPrivate` grey — a vehicle type this build does not know is still drawn,
    /// because a missing marker reads as a platform outage and an unfamiliar grey one reads as what
    /// it is.
    static func markerColourExpression(palette: VehiclePalette) -> NSExpression {
        var stops: [NSExpression: NSExpression] = [:]
        for (wire, colour) in palette.legend {
            stops[NSExpression(forConstantValue: wire)] = NSExpression(forConstantValue: colour)
        }
        return NSExpression(
            forMLNMatchingKey: NSExpression(forKeyPath: propVehicleType),
            in: stops,
            default: NSExpression(forConstantValue: palette.fallback)
        )
    }

    static let clusterMaxZoom = 14
    static let clusterRadiusPx = 60
    private static let clusterRadiusSmall: CGFloat = 16
    private static let clusterRadiusMedium: CGFloat = 22
    private static let clusterRadiusLarge: CGFloat = 28
    private static let clusterStepMedium = 10
    private static let clusterStepLarge = 40
    private static let clusterTextSize: CGFloat = 12
    private static let accuracyOpacity: CGFloat = 0.15
    private static let geofenceOpacity: CGFloat = 0.08
    private static let geofenceStrokePx: CGFloat = 2
    private static let routeWidthPx: CGFloat = 5
    private static let routeCasingWidthPx: CGFloat = 9

    /// Narrower than the route: a walking leg is a hint, not the journey.
    private static let walkWidthPx: CGFloat = 4
    private static let walkDash: CGFloat = 2
    private static let walkGap: CGFloat = 2
    private static let pinRadiusPx: CGFloat = 7
    private static let pinStrokePx: CGFloat = 2

    /// How faded a last-known marker is (SCR-PI-032). Legible, and unmistakably not live.
    static let staleOpacity: CGFloat = 0.45
}
